ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java

```diff
@@ -5208,55 +5208,94 @@ private static void moveAcidFiles(String deltaFileType, PathFilter pathFilter, F
+        try {
+          future.get();
+        } catch (Exception e) {
+          throw handlePoolException(pool, e);
         }
-        FileStatus[] bucketStats = fs.listStatus(deltaPath, AcidUtils.bucketFileFilter);
-        LOG.debug("Acid move found " + bucketStats.length + " bucket files");
-        for (FileStatus bucketStat : bucketStats) {
-          Path bucketSrc = bucketStat.getPath();
-          Path bucketDest = new Path(deltaDest, bucketSrc.getName());
-          final String msg = "Unable to move source " + bucketSrc + " to destination " +
-              bucketDest;
-          LOG.info("Moving bucket " + bucketSrc.toUri().toString() + " to " +
-              bucketDest.toUri().toString());
-          try {
-            fs.rename(bucketSrc, bucketDest);
-            if (newFiles != null) {
-              newFiles.add(bucketDest);
+      }
+    }
+  }
+
+  private static void moveAcidFilesForDelta(String deltaFileType, FileSystem fs,
+                                            Path dst, Set<Path> createdDeltaDirs,
+                                            List<Path> newFiles, FileStatus deltaStat) throws HiveException {
+
+    Path deltaPath = deltaStat.getPath();
+    // Create the delta directory.  Don't worry if it already exists,
+    // as that likely means another task got to it first.  Then move each of the buckets.
+    // it would be more efficient to try to move the delta with it's buckets but that is
+    // harder to make race condition proof.
+    Path deltaDest = new Path(dst, deltaPath.getName());
+    try {
+      if (!createdDeltaDirs.contains(deltaDest)) {
+        try {
+          if(fs.mkdirs(deltaDest)) {
+            try {
+              fs.rename(AcidUtils.OrcAcidVersion.getVersionFilePath(deltaStat.getPath()),
+                      AcidUtils.OrcAcidVersion.getVersionFilePath(deltaDest));
+            } catch (FileNotFoundException fnf) {
+              // There might be no side file. Skip in this case.
             }
-          } catch (Exception e) {
-            throw getHiveException(e, msg);
           }
+          createdDeltaDirs.add(deltaDest);
+        } catch (IOException swallowIt) {
+          // Don't worry about this, as it likely just means it's already been created.
+          LOG.info("Unable to create " + deltaFileType + " directory " + deltaDest +
+                  ", assuming it already exists: " + swallowIt.getMessage());
         }
-      } catch (IOException e) {
-        throw new HiveException("Error moving acid files " + e.getMessage(), e);
       }
+      FileStatus[] bucketStats = fs.listStatus(deltaPath, AcidUtils.bucketFileFilter);
+      LOG.debug("Acid move found " + bucketStats.length + " bucket files");
+      for (FileStatus bucketStat : bucketStats) {
+        Path bucketSrc = bucketStat.getPath();
+        Path bucketDest = new Path(deltaDest, bucketSrc.getName());
+        final String msg = "Unable to move source " + bucketSrc + " to destination " +
```

Comment:
Minor comment, can we move this to catch block?

Reply:
Done